Remote signing client/server.

cheatfate opened the pull request:
This PR adds support for https://github.com/ConsenSys/web3signer remote signature server. From my point of view `web3signer` is the most full version of remote signer, because EIP-3030 (https://eips.ethereum.org/EIPS/eip-3030) do not support enough entities to run even on `phase0` network.

To use remote signature for one of the validator's public keys you need to create new keystore file inside `<data>/validators/<public_key>` directory.

This file should be placed in `validators/<public key>/remote_keystore.json` file.

While loading keys, `nimbus_beacon_node` will follow this rules.
First it will try to load local validator keystore `keystore.json`. If it not accessible
it will try to load `remote_keystore.json` file. Only if both attempts filed it will show an error.

#### Example

```json
{
  "version": "1",
  "description": "This is simple remote keystore file",
  "type": "web3signer",
  "pubkey": "0x8107ff6a5cfd1993f0dc19a6a9ec7dc742a528dd6f2e3e10189a4a6fc489ae6c7ba9070ea4e2e328f0d20b91cc129733",
  "remote": "http://127.0.0.1:15052",
  "ignore_ssl_verification": true
}
```

#### Fields

1. `version` - optional string encoded integer.
2. `description` - optional string encoded description.
3. `type` - optional (by default `web3signer`) string encoded type of remote
signer. Right now only `web3signer` value is supported, but maybe in future we are going to support `dirk`.
4. `pubkey` - required hexadecimal string of validator's public key.
5. `remote` - required URL of remote service.
6. `ignore_ssl_verification` - optional boolean flag which provides option to  ignore self-signed certificates.

cheatfate committed:
Initial commit.

cheatfate committed:
Use new method for SignRequest deserialization.

cheatfate committed:
Missed file.

cheatfate committed:
Add block signatures.

cheatfate committed:
Switch from EIP-3030 to Web3Signer specification.

cheatfate committed:
Add forgotten import file.

cheatfate committed:
Add signing node to Makefile.
Add request-timeout option.
Remove old signing process.

cheatfate committed:
One more fix test.

cheatfate committed:
Fix compilation issues in signing node.
Move REST constants to different file to avoid problems with json-rpc.

cheatfate committed:
Fix generic sandwich.

cheatfate committed:
Attempt to fix one more generic sandwich.

cheatfate committed:
One more attempt to fix sandwich in jenkins.

cheatfate committed:
Fix linux sandwich.

cheatfate committed:
Fix local_testnet script.

cheatfate committed:
Add metrics and delay calculation.

cheatfate committed:
Fix one more windows sandwich.

cheatfate committed:
Post rebase fixes.

cheatfate committed:
Fix gossip test.

cheatfate committed:
Proper fix for gossip test.

cheatfate committed:
Attempt to fix one more sandwich on linux.

cheatfate committed:
Add new-stew fix branch.

cheatfate committed:
Bump stew.

cheatfate committed:
Fix imports.

zah commented on beacon_chain/spec/keystore.nim:
```diff
@@ -497,6 +512,75 @@ proc readValue*(r: var JsonReader, value: var Kdf)
     r.raiseUnexpectedValue(
       "The Kdf value should have sub-fields named 'function' and 'params'")
 
+proc readValue*(r: var JsonReader, value: var RemoteKeystore)
```
Why do we need to implement a custom serialization for this type? Isn't the definition something that we control? Why is the default implementation not adequate?

zah commented on beacon_chain/conf.nim:
```diff
@@ -33,6 +33,7 @@ const
   # Maybe there should be a config option for this.
   defaultListenAddress* = (static ValidIpAddress.init("0.0.0.0"))
   defaultAdminListenAddress* = (static ValidIpAddress.init("127.0.0.1"))
+  DefaultSigningNodeRequestTimeout* = 60
```
We tend to use `camelCase` for constants (just like the lines above). `PascalCase` is reserved for types or compile-time variables storing types.

zah commented on beacon_chain/conf.nim:
```diff
@@ -590,7 +594,84 @@ type
+      defaultValue: DefaultSigningNodeRequestTimeout
+      name: "request-timeout" }: int
+
+    case cmd* {.
```
Do you anticipate any commands to be added in the future?

cheatfate replied:
I'm not decided what tool should be responsible to help user with creation of `remote_keystore.json` files.

zah commented on beacon_chain/nimbus_signing_node.nim:
```diff
@@ -0,0 +1,355 @@
+    case item.kind
+    of ValidatorKind.Local:
+      let pubkey = item.privateKey.toPubKey().toPubKey()
+      if pubkey in duplicates:
```
You can check whether the validator was already added with the `getValidator` function of the `ValidartorPool` (`sn.attachedValidators`). This will result in a linear algorithms instead of the current quadratic one which is not that nice considering that we run servers with thousands of validators.

zah commented on beacon_chain/spec/eth2_apis/eth2_rest_serialization.nim:
```diff
@@ -319,6 +324,14 @@ proc sszResponse*(t: typedesc[RestApiResponse], data: auto): RestApiResponse =
+                          requireAllFields = true): Result[T, cstring] =
+  try:
+    ok(RestJson.decode(string(data), T, requireAllFields = requireAllFields))
+  except SerializationError:
```
The `SerializationError` exception carries useful diagnostic information (i.e. line and column information for the unexpected syntax/values in the json string). Why is this helper considered an improvement if it destroys the diagnostic information?

cheatfate replied:
Its not an improvement its the way how to avoid exceptions and code duplication in object parser's code.

zah commented on beacon_chain/spec/eth2_apis/eth2_rest_serialization.nim:
```diff
@@ -945,6 +958,355 @@ proc readValue*(reader: var JsonReader[RestJson],
     reader.raiseUnexpectedValue($res.error())
 
+# Web3SignerRequest
+proc writeValue*(writer: var JsonWriter[RestJson],
```
`nim-json-serialization` already supports case objects. Why is this definition necessary?

cheatfate replied:
`nim-json-serialization` supports case objects only when `case` delimiter is specified before other fields, it means that JSON should be formed in some strict way before we will be able to use it.

zah commented on beacon_chain/spec/eth2_apis/eth2_rest_serialization.nim:
```diff
@@ -945,6 +958,355 @@ proc readValue*(reader: var JsonReader[RestJson],
+
+  for fieldName in readObjectFields(reader):
+    case fieldName
+    of "type":
```
Is any of the web3signer implementations not listing the `type` field as a first field in the `Web3SignerRequest`? nim-json-serialization already supports case objects and should be able to handle the decoding otherwise.

cheatfate replied:
@zah This is current specification https://consensys.github.io/web3signer/web3signer-eth2.html and i dont see any requirements for field `type` to be first, This is JSON, so `type` field could first or could be last. That's actually why i need to implement this custom type deserialization.

zah commented on beacon_chain/spec/eth2_apis/rest_remote_signer_calls.nim:
```diff
@@ -0,0 +1,184 @@
+                  $res.error() & "]"
+        inc(nbc_remote_signer_failures)
+        return Web3SignerDataResponse.err(msg)
+      let sig = res.get().signature.load()
```
I think it makes sense to verify that the signature is valid here (against the public key and the data) before we broadcast the message on the network. 

I can imagine some configuration issues that may result in invalid signatures. If this happens, we don't want to get de-scored and kicked out from the network if only some of the validators have invalid configuration. Furthermore, we need to inform the user for the problem instead of failing silently.

zah commented on beacon_chain/validators/keystore_management.nim:
```diff
@@ -312,8 +328,29 @@ proc loadKeystoreUnsafe*(validatorsDir, secretsDir,
+        except SerializationError as e:
+          error "Failed to read remote keystore file",
+                keystore_path = remoteKeystorePath,
+                err_msg = e.formatMsg("<remote_keystore>")
```
```suggestion
                err_msg = e.formatMsg(remoteKeystorePath)
```

The exception carries only line and column information (to reduce the size of intermediate objects and to avoid string allocations). The argument to `formatMsg` is the filename that will be printed together with the line and column error.

zah commented on beacon_chain/validators/validator_duties.nim:
```diff
@@ -86,37 +86,56 @@ proc findValidator(validators: auto, pubKey: ValidatorPubKey):
-    findValidator(validators, pubKey.toPubKey()))
+  let
+    pubKey = item.privateKey.toPubKey()
+    index = findValidator(validators, pubKey.toPubKey())
```
It's worth checking whether these linear searches contribute to poor start-up times of our Prater fleet nodes. We can populate a temporary index with a hash table to get a linear algorithm.

cheatfate replied:
@zah this is not a new code it was present there before my PR. Even if this linear search over 100k validators could take some time in milliseconds, the biggest problem is how keystores are encoded/decoded (number of iterations) seriously affects our performance.

zah commented on beacon_chain/validator_client/duties_service.nim:
```diff
@@ -152,20 +152,30 @@ proc pollForAttesterDuties*(vc: ValidatorClientRef,
+    var pending: seq[Future[SignatureResult]]
+    var validators: seq[AttachedValidator]
     for item in addOrReplaceItems:
       let validator = vc.attachedValidators.getValidator(item.duty.pubkey)
```
Is there a guarantee that `getValidator` cannot return `nil` here?

cheatfate replied:
@zah if you check line 125 you will see

```nim
    relevantDuties = duties.filterIt(
      checkDuty(it) and (it.pubkey in vc.attachedValidators)
    )
```
It means that `relevantDuties` will hold only duties with validators we know.

Also you could check that `addOrReplaceItems` is just a subset of `relevantDuties`, so `getValidator` could not return `nil` here.

zah commented on beacon_chain/validators/validator_duties.nim:
```diff
@@ -86,37 +86,56 @@ proc findValidator(validators: auto, pubKey: ValidatorPubKey):
+    warn "Unable to resolve remote signer address",
+         remote_url = $item.remoteUrl, validator = item.publicKey
+    return
+  let index = findValidator(validators, item.publicKey.toPubKey())
```
Ditto

cheatfate replied:
Answer is above.

cheatfate committed:
Fix constant case.
Remove duplicates checks for filesystem keystores.
Add remoteKeystorePath in error.
Fix test.

cheatfate committed:
Fix constant in nimbus_signing_node.
